Fix aot sequence_nr to reset bwd flag

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107210

The way the aot autograd sequence_nr tracking works is that we run the aot export logic, the dynamo captured forward graph is run under an fx.Interpreter, which iterates through the nodes of the forward graph while setting the `current_metadata`. 
Since during backward what is run doesn't correspond to any node during forward, we fallback to the global `current_metadata`. And since this global metadata is ends up being shared between runs, that leads to weirdness if we forget to reset things, e.g., depending whether this is the first test run, the printed results will be different.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

soulitzer committed:
Fix aot sequence_nr to reset bwd flag

[ghstack-poisoned]

albanD commented:
Will let Brian comment on this.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -166,7 +166,7 @@ def prehook(grad_output):
         def posthook(grad_input, grad_output):
             fx_traceback.set_stack_trace(special_stack_)
-            fx_traceback.set_seq_nr(-1, bwd=True)
+            fx_traceback.set_seq_nr(-1, bwd=False)
```
> Since during backward what is run doesn't correspond to any node during forward, we fallback to the global current_metadata. And since this global metadata is ends up being shared between runs, that leads to weirdness if we forget to reset things, e.g., depending whether this is the first test run, the printed results will be different.

Could you explain how the description maps to this bugfix? It doesn't seem like this will reset anything (just indicate to anyone querying the currently executed node's metadata that we are not currently executing a backward node). But I'm probably missing something.

soulitzer replied:
> Indicate to anyone querying the currently executed node's metadata that we are not currently executing a backward node)

Yup that is correct, we register prehooks and posthooks so that during backward, whether we are currently executing a backward node would be reflected on the current_node's metadata. 

And yeah, we don't actually completely reset its value to "unset" we just leave it as `in_bwd=False`, but that seems okay to me because at least in the place where we currently query it, those two is treated the same, i.e. we do `current_node.get("in_bwd", False)`.

The problem arises when we do forward/backward the second time, but in a subtle way. When we do forward, the ops that we execute in forward map to ops that are in the fx graph passed to the Interpreter, so we'd have a fresh current_node to work with, so those are totally fine. 

The weirdness happens when we reach something like `ones_like` which is in a weird position because technically is not quite part of the forward, but not part of any backward node's execution either. For the purposes of aot sequence_nr though we'd like it to be considered part of the forward. 

Today without this bug fix it can be sometimes counted as backward because `in_bwd` is not reset from the previous backward.

bdhirsh replied:
Thanks for the explanation!

Now I see - looks like you're right that if we're not in the middle of executing a node, then `current_meta` (here) will be some global dictionary. Since we aren't in the middle of executing a node while the autograd engine is running the backward, then these pre and post hooks end up polluting the global dictionary.


It seems like morally what we want is for the pre_hook to temporarily add `seq_nr=K, bwd=True` to the global dict, and for the post_hook to do the cleanup on exit (reverting to whatever the previous value in `curr_meta` was, or scrubbing those properties from the dictionary if they weren't set).

Setting them to -1 and False on exit seems... reasonable enough. What do you think of trying to set to the old state? (Having the pre_hook save the old state, so the post_hook can reset it)

soulitzer replied:
Yeah doing that sounds reasonable as well. Though we don't current plan to support reentrant autograd, this would be helpful toward that if we choose to do so in the future. I've updated this PR to do that.

soulitzer committed:
Update on "Fix aot sequence_nr to reset bwd flag"


The way the aot autograd sequence_nr tracking works is that we run the aot export logic, the dynamo captured forward graph is run under an fx.Interpreter, which iterates through the nodes of the forward graph while setting the `current_metadata`. 
Since during backward what is run doesn't correspond to any node during forward, we fallback to the global `current_metadata`. And since this global metadata is ends up being shared between runs, that leads to weirdness if we forget to reset things, e.g., depending whether this is the first test run, the printed results will be different.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

soulitzer committed:
Update on "Fix aot sequence_nr to reset bwd flag"


The way the aot autograd sequence_nr tracking works is that we run the aot export logic, the dynamo captured forward graph is run under an fx.Interpreter, which iterates through the nodes of the forward graph while setting the `current_metadata`. 
Since during backward what is run doesn't correspond to any node during forward, we fallback to the global `current_metadata`. And since this global metadata is ends up being shared between runs, that leads to weirdness if we forget to reset things, e.g., depending whether this is the first test run, the printed results will be different.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer commented:
@pytorchbot merge -f "Unrelated"

soulitzer commented:
@pytorchbot merge -f "Unrelated failures"

soulitzer committed:
Update on "Fix aot sequence_nr to reset bwd flag"


The way the aot autograd sequence_nr tracking works is that we run the aot export logic, the dynamo captured forward graph is run under an fx.Interpreter, which iterates through the nodes of the forward graph while setting the `current_metadata`. 
Since during backward what is run doesn't correspond to any node during forward, we fallback to the global `current_metadata`. And since this global metadata is ends up being shared between runs, that leads to weirdness if we forget to reset things, e.g., depending whether this is the first test run, the printed results will be different.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge
